### PR TITLE
FIX-#719:  zip_iterator for relaxed iterators. Up to missmatch prototype

### PR DESCRIPTION
**Pull request by DenisYaroshevskiy**
Got `zip_iterator` to the point where mismatch is visible and actually wrote a prototype for it.
Everything that has to do with `preprocess_range` for `zip_iterator` is so far a complete nonsense, I just wanted it to work.

List of changes:
* Removed `forward_to_unaligned`. Seems to break.
* `unalign` free function. Just a tiny helper.
* `++/--` in `operations_with_distance`. zip_iterator tests cover these.
* `zip_iterator` will now work with practically anything. The enablement of `eve::algo::iterator` functions will only happen if the components actually support it.
* wrote a prototyped `preprocess_range` for `zip_iterator` only for case when the sentinel is for the first iterator in a group.
* tested a prototype missmatch.

The next step will be to do `zip` with proper `preprocess_range` that supports all the necessary things.

### CODE REVIEW

File: include/eve/algo/ptr_iterator.hpp [hunk 1]
```diff
@@ -38,7 +38,7 @@ namespace eve::algo
     template <typename _Cardinal>
     auto cardinal_cast(_Cardinal) const { return unaligned_ptr_iterator<T, _Cardinal>{ptr}; }
 
-    value_type& operator*() const { return *ptr; }
+    auto& operator*() const { return *ptr; }
```
Comment (jfalcou) [inline include/eve/algo/ptr_iterator.hpp:41]: why this change ?

**Review by jfalcou — APPROVED** [review]

File: include/eve/algo/ptr_iterator.hpp (hunk 1 above)
Comment (DenisYaroshevskiy) [reply include/eve/algo/ptr_iterator.hpp:41]: Because `value_type` for `const int*` is `int` but the reference type is `const int&`
